src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkUtils.java

```diff
@@ -441,4 +502,54 @@ private static BigQueryFileFormat getFileFormat(List<BigQueryTableFieldSchema> f
     }
     return AvroOutputFormat.class;
   }
+
+  public static String generateUpdateUpsertQuery(Operation operation,
+                                                 TableId sourceTableId,
+                                                 TableId destinationTableId,
+                                                 List<String> tableFieldsList,
+                                                 List<String> tableKeyList,
+                                                 List<String> orderedByList,
+                                                 String partitionFilter) {
+
+    String source = String.format("`%s.%s.%s`",
+                                  sourceTableId.getProject(),
+                                  sourceTableId.getDataset(),
+                                  sourceTableId.getTable());
+    String destination = String.format("`%s.%s.%s`",
+                                       destinationTableId.getProject(),
+                                       destinationTableId.getDataset(),
+                                       destinationTableId.getTable());
+
+    String criteria = tableKeyList.stream().map(s -> String.format(CRITERIA_TEMPLATE, s, s))
+      .collect(Collectors.joining(" AND "));
+    criteria = partitionFilter != null ? String.format("(%s) AND %s",
+                                                       formatPartitionFilter(partitionFilter), criteria) : criteria;
+    String fieldsForUpdate = tableFieldsList.stream().filter(s -> !tableKeyList.contains(s))
+      .map(s -> String.format(CRITERIA_TEMPLATE, s, s)).collect(Collectors.joining(", "));
+    String orderedBy = orderedByList.isEmpty() ? "" : " ORDER BY " + String.join(", ", orderedByList);
+    String sourceTable = String.format(SOURCE_DATA_QUERY, String.join(", ", tableKeyList), orderedBy, source);
+    switch (operation) {
+      case UPDATE:
+        return String.format(UPDATE_QUERY, destination, fieldsForUpdate, sourceTable, criteria);
+      case UPSERT:
+        String insertFields = String.join(", ", tableFieldsList);
+        return String.format(UPSERT_QUERY, destination, sourceTable, criteria, fieldsForUpdate,
+                             insertFields, insertFields);
+      default:
+        return "";
+    }
+  }
+
+  private static String formatPartitionFilter(String partitionFilter) {
+    String[] queryWords = partitionFilter.split(" ");
+    int index = 0;
+    for (String word: queryWords) {
+      if (COMPARISON_OPERATORS.contains(word.toUpperCase())) {
+        queryWords[index - 1] = queryWords[index - 1].replace(queryWords[index - 1],
```

Comment:
In the first iteration, if the condition is true then this will cause ArrayIndexOutOfBoundsException?

Reply:
This means the expression would be invalid. A partition filter is a boolean expression so it has to be something like:

'a < b' or 'x = y' and so on.

Since we are only looking for the comparison operators when evaluating, we are guaranteed to have a "previous" element.

Reply:
I think it's more about "invalid" query, e.g. if anyone tries to start expression with "<" this would fail with ArrayIndexOutOfBoundsException instead of a nice one